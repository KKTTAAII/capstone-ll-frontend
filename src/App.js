import React, { useEffect } from "react";
import Routes from "./Routes/Routes";
import PetlyApi from "./api";
import UserInfoContext from "./common/UserInfoContext";
import { BrowserRouter, Redirect } from "react-router-dom";
import { useLocalStorageState } from "./common/hooks";

import jwt_decode from "jwt-decode";
import swal from "sweetalert";

const App = () => {
  const [token, setToken] = useLocalStorageState("token", null);
  const [user, setUser] = useLocalStorageState("user", null);

  useEffect(() => {
    async function getUser() {
      try {
        //check first if there is token in localStorage
        if (token) {
          //decode the token
          const tokenInfo = jwt_decode(token);
          //check the userType on token because we have two types: adopters and shelters
          if (tokenInfo.userType === "shelters") {
            const user = await PetlyApi.get(
              tokenInfo.userType,
              tokenInfo.id,
              token
            );
            //check if we have this user in db, if not, token may expire or no that user
            if (user) {
              setUser(tokenInfo);
            }
          } else if (tokenInfo.userType === "adopters") {
            const user = await PetlyApi.get(
              tokenInfo.userType,
              tokenInfo.username,
              token
            );
            //check if we have this user in db, if not, token may expire or no that user
            if (user) {
              setUser(tokenInfo);
            }
          }
        }
      } catch (err) {
        setUser(null);
        swal(err[0]);
        console.log(err);
        return <Redirect to="/" />;
      }
    }
    getUser();
  }, [token]);

  const signUp = async (userType, data) => {
    try {
      const response = await PetlyApi.register(userType, data);
      const user = jwt_decode(response);
      setToken(response);
      setUser(user);
    } catch (err) {
      console.log(err);
      return err;
    }
  };

  const logIn = async (userType, data) => {
    try {
      const response = await PetlyApi.logIn(userType, data);
      const user = jwt_decode(response);
      setToken(response);
      setUser(user);
    } catch (err) {
      console.log(err);
      return err;
    }
  };

  const logOut = () => {
    setToken(null);
    setUser(null);
  };
  return (
    <BrowserRouter>
      <UserInfoContext.Provider value={{ user, token }}>
        <div className="App">
          <Routes signUp={signUp} logIn={logIn} />
        </div>
      </UserInfoContext.Provider>
    </BrowserRouter>
  );
};

export default App;
