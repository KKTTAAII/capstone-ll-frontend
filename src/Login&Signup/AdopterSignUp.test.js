import React from "react";
import {render} from "@testing-library/react";
import AdopterSignUp from "./AdopterSignUp.js";
import {MemoryRouter} from "react-router-dom";
import App from "../App";

jest.mock("../App");
App.signUp = jest.fn();

it("renders without crashing", () => {
  render(
    <MemoryRouter>
      <AdopterSignUp signUp={App.signUp} />
    </MemoryRouter>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <AdopterSignUp signUp={App.signUp} />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
