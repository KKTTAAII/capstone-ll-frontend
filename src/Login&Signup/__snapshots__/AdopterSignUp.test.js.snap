// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="AdopterSignup-container"
  >
    <div
      class="AdopterSignup-header"
    >
      Adopter Sign up
    </div>
    <form
      class="AdopterSignup-form"
    >
      <label
        class="AdopterSignup-label"
      >
        Username*:
      </label>
      <input
        class="AdopterSignup-input"
        id="username"
        name="username"
        placeholder="Username"
        type="text"
        value=""
      />
      <label
        class="AdopterSignup-label"
      >
        Password*:
      </label>
      <input
        class="AdopterSignup-input"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <label
        class="AdopterSignup-label"
      >
        Email*:
      </label>
      <input
        class="AdopterSignup-input"
        id="email"
        name="email"
        placeholder="Email"
        type="email"
        value=""
      />
      <div
        class="AdopterSignup-preview-img-container"
      >
        <img
          alt=""
          class="AdopterSignup-preview-img"
          src=""
        />
        <div>
          Profile Image
        </div>
        <input
          class="AdopterSignup-img-button"
          id="picture"
          name="picture"
          type="file"
        />
        <label
          for="picture"
        >
          <div
            class="AdopterSignup-add-picture-button"
          >
            Add picture
          </div>
        </label>
      </div>
      <label
        class="AdopterSignup-label"
        for="privateOutdoors"
      >
        Private Outdoors:
      </label>
      <select
        class="AdopterSignup-select"
        id="privateOutdoors"
        name="privateOutdoors"
      >
        <option
          value="0"
        >
          No
        </option>
        <option
          value="1"
        >
          Yes
        </option>
      </select>
      <label
        class="AdopterSignup-label"
        for="numOfDogs"
      >
        Number of dogs:
      </label>
      <input
        class="AdopterSignup-select"
        id="numOfDogs"
        max="20"
        min="0"
        name="numOfDogs"
        type="number"
        value="0"
      />
      <label
        class="AdopterSignup-label"
        for="preferredGender"
      >
        Preferred Gender:
      </label>
      <select
        class="AdopterSignup-select"
        id="preferredGender"
        name="preferredGender"
      >
        <option
          value="Female"
        >
          Female
        </option>
        <option
          value="Male"
        >
          Male
        </option>
      </select>
      <label
        class="AdopterSignup-label"
        for="preferredAge"
      >
        Preferred Age:
      </label>
      <select
        class="AdopterSignup-select"
        id="preferredAge"
        name="preferredAge"
      >
        <option
          value="Baby"
        >
          Baby
        </option>
        <option
          value="Young"
        >
          Young
        </option>
        <option
          value="Adult"
        >
          Adult
        </option>
        <option
          value="Senior"
        >
          Senior
        </option>
      </select>
      <label
        class="AdopterSignup-label"
        for="description"
      >
        Adopter's Bio:
      </label>
      <textarea
        cols="45"
        id="description"
        name="description"
        rows="5"
      />
      <button
        class="AdopterSignup-button"
      >
        Sign up
      </button>
    </form>
    <div
      class="AdopterSignup-alreadySignup"
    >
      Already signed up? 
      <a
        class="AdopterSignup-signupLink"
        href="/adopters/login"
      >
        Log in
      </a>
    </div>
  </div>
</DocumentFragment>
`;
